File: bindings/go/satox_bindings/api_manager.go
package satox_bindings

import (
	"errors"
	"fmt"
	"net/http"
	"sync"
)

// APIManager provides API server management and endpoint registration.
type APIManager struct {
	initialized   bool
	serverRunning bool
	host          string
	port          int
	server        *http.Server
	endpoints     map[string]http.HandlerFunc
	mu            sync.RWMutex
}

// NewAPIManager creates a new instance of APIManager.
func NewAPIManager() *APIManager {
	return &APIManager{
		endpoints: make(map[string]http.HandlerFunc),
	}
}

// Initialize initializes the API manager.
func (m *APIManager) Initialize() error {
	m.mu.Lock()
	defer m.mu.Unlock()
	if m.initialized {
		return errors.New("APIManager already initialized")
	}
	m.initialized = true
	return nil
}

// StartServer starts the API server on the specified host and port.
func (m *APIManager) StartServer(host string, port int) error {
	m.mu.Lock()
	defer m.mu.Unlock()
	if !m.initialized {
		return errors.New("APIManager not initialized")
	}
	if m.serverRunning {
		return errors.New("API server already running")
	}

	mux := http.NewServeMux()
	for path, handler := range m.endpoints {
		mux.HandleFunc(path, handler)
	}

	m.host = host
	m.port = port
	m.server = &http.Server{
		Addr:    fmt.Sprintf("%s:%d", host, port),
		Handler: mux,
	}

	m.serverRunning = true
	go func() {
		_ = m.server.ListenAndServe()
	}()
	return nil
}

// StopServer stops the API server.
func (m *APIManager) StopServer() error {
	m.mu.Lock()
	defer m.mu.Unlock()
	if !m.serverRunning {
		return errors.New("API server not running")
	}
	if m.server != nil {
		err := m.server.Close()
		m.serverRunning = false
		m.server = nil
		return err
	}
	m.serverRunning = false
	return nil
}

// IsServerRunning returns whether the API server is running.
func (m *APIManager) IsServerRunning() bool {
	m.mu.RLock()
	defer m.mu.RUnlock()
	return m.serverRunning
}

// RegisterEndpoint registers a new API endpoint.
func (m *APIManager) RegisterEndpoint(path string, handler http.HandlerFunc) error {
	m.mu.Lock()
	defer m.mu.Unlock()
	if !m.initialized {
		return errors.New("APIManager not initialized")
	}
	if _, exists := m.endpoints[path]; exists {
		return errors.New("Endpoint already registered")
	}
	m.endpoints[path] = handler
	return nil
}

// UnregisterEndpoint removes an API endpoint.
func (m *APIManager) UnregisterEndpoint(path string) error {
	m.mu.Lock()
	defer m.mu.Unlock()
	if !m.initialized {
		return errors.New("APIManager not initialized")
	}
	if _, exists := m.endpoints[path]; !exists {
		return errors.New("Endpoint not found")
	}
	delete(m.endpoints, path)
	return nil
}

// GetServerStatus returns the current server status.
func (m *APIManager) GetServerStatus() map[string]interface{} {
	m.mu.RLock()
	defer m.mu.RUnlock()
	return map[string]interface{}{
		"running": m.serverRunning,
		"host":    m.host,
		"port":    m.port,
	}
}

// Shutdown shuts down the API manager and server.
func (m *APIManager) Shutdown() error {
	m.mu.Lock()
	defer m.mu.Unlock()
	if m.serverRunning && m.server != nil {
		_ = m.server.Close()
		m.serverRunning = false
		m.server = nil
	}
	m.initialized = false
	m.endpoints = make(map[string]http.HandlerFunc)
	return nil
}

File: bindings/java/satox_bindings/APIManager.java
package com.satox.bindings;

import java.util.Map;
import java.util.concurrent.locks.ReadWriteLock;
import java.util.concurrent.locks.ReentrantReadWriteLock;

public class APIManager {
    private final ReadWriteLock lock;
    private boolean initialized;
    private boolean serverRunning;
    private String host;
    private int port;
    private boolean sslEnabled;
    private String sslCertPath;
    private String sslKeyPath;

    public APIManager() {
        this.lock = new ReentrantReadWriteLock();
        this.initialized = false;
        this.serverRunning = false;
    }

    public void initialize() {
        lock.writeLock().lock();
        try {
            if (initialized) {
                throw new IllegalStateException("APIManager already initialized");
            }
            initialized = true;
        } finally {
            lock.writeLock().unlock();
        }
    }

    public void startServer(String host, int port, boolean enableSSL, String sslCertPath, String sslKeyPath) {
        ensureInitialized();
        lock.writeLock().lock();
        try {
            if (serverRunning) {
                throw new IllegalStateException("Server is already running");
            }
            this.host = host;
            this.port = port;
            this.sslEnabled = enableSSL;
            this.sslCertPath = sslCertPath;
            this.sslKeyPath = sslKeyPath;
            this.serverRunning = true;
        } finally {
            lock.writeLock().unlock();
        }
    }

    public void stopServer() {
        ensureInitialized();
        lock.writeLock().lock();
        try {
            if (!serverRunning) {
                throw new IllegalStateException("Server is not running");
            }
            this.serverRunning = false;
        } finally {
            lock.writeLock().unlock();
        }
    }

    public boolean isServerRunning() {
        lock.readLock().lock();
        try {
            return serverRunning;
        } finally {
            lock.readLock().unlock();
        }
    }

    public void registerEndpoint(String path, String method, Object handler) {
        ensureInitialized();
        ensureServerRunning();
        // Implementation for registering API endpoints
    }

    public void unregisterEndpoint(String path, String method) {
        ensureInitialized();
        ensureServerRunning();
        // Implementation for unregistering API endpoints
    }

    public void setMiddleware(Object middleware) {
        ensureInitialized();
        ensureServerRunning();
        // Implementation for setting API middleware
    }

    public Map<String, Object> getServerStatus() {
        ensureInitialized();
        lock.readLock().lock();
        try {
            return Map.of(
                "running", serverRunning,
                "host", host,
                "port", port,
                "ssl_enabled", sslEnabled
            );
        } finally {
            lock.readLock().unlock();
        }
    }

    public void shutdown() {
        lock.writeLock().lock();
        try {
            if (serverRunning) {
                stopServer();
            }
            initialized = false;
        } finally {
            lock.writeLock().unlock();
        }
    }

    private void ensureInitialized() {
        if (!initialized) {
            throw new IllegalStateException("APIManager not initialized");
        }
    }

    private void ensureServerRunning() {
        if (!serverRunning) {
            throw new IllegalStateException("Server is not running");
        }
    }
} 